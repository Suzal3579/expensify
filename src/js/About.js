import React from 'react';
import { Header, List } from 'semantic-ui-react';

const About = () => (
    <div>
        <Header color="grey">About</Header>
        <p>
            So, some of your friends/ relatives owe you some money? <br />
            If yes, then you are not alone. This happens to every one of us.
            We lend people some money and later forget to take it back. This app
            takes information about your money and even keeps a date including
            some handy notes to remember later. Now all you have to do is remember
            to open this app and all your money details/ expenditure are available
            right on your screen.
        </p>
        <Header color="grey">Developer</Header>
        <List>
            <List.Item icon="users" content="Pramesh Bajracharya" />
            <List.Item icon="marker" content="Kathmandu, Nepal" />
            <List.Item icon="mail" content={<a href="mailto:pe.messh@gmail.com">pe.messh@gmail.com</a>} />
            <List.Item icon="linkify" content={<a href="http://bajracharya.me">My website and blogs.</a>} />
        </List>
    </div>
);

export default About;
