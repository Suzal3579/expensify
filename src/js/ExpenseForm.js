import React, { Component } from 'react';
import 'react-dates/initialize';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? props.expense.amount : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            errorState: '',
        };
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    }
    onDateChange = (createdAt) => {
        if (createdAt) { this.setState(() => ({ createdAt })); }
    }
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    }
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({
                errorState: 'Sorry but you cannot leave these fields empty ...',
            }));
        } else {
            this.setState(() => ({
                errorState: '',
            }));
            this.props.onSubmit({
                description: this.state.description,
                note: this.state.note,
                amount: this.state.amount,
                createdAt: this.state.createdAt.valueOf(),
            });
        }
    }
    render() {
        return (
            <div>
                {this.state.errorState && <p>{this.state.errorState}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        value={this.state.description}
                        placeholder="Description"
                        onChange={this.onDescriptionChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <br />
                    <input
                        type="text"
                        value={this.state.note}
                        placeholder="Note"
                        onChange={this.onNoteChange}
                    />
                    <input
                        type="text"
                        value={this.state.amount}
                        placeholder="Amount"
                        onChange={this.onAmountChange}
                    />
                    <button type="submit">Add</button>
                </form>
            </div>
        );
    }
}

