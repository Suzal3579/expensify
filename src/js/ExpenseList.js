import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Header } from 'semantic-ui-react';

import { database } from '../firebase/firebase';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from '../apis/getVisibleExpenses';

export class ExpenseList extends Component {
    constructor(props) {
        super(props);
        this.state = {
            expenses: props.expenses,
        };
    }
    componentWillMount() {
        database.ref('expenses')
            .on('value', (snapshot) => {
                const dataReceived = snapshot.val();
                const expensesArrayFromFirebase =
                    Object.keys(dataReceived)
                        .map(a => [
                            a,
                            dataReceived[a].description,
                            dataReceived[a].note,
                            dataReceived[a].createdAt,
                            dataReceived[a].amount]);
                this.setState(() => ({
                    expenses: expensesArrayFromFirebase,
                }));
            });
    }
    render() {
        return (
            <div className="row">
                <hr />
                {
                    this.state.expenses.length === 0 ? (
                        <Header as="h2" color="grey" textAlign="center">No expenses</Header>
                    ) : (
                        <ExpenseListItem expenses={this.state.expenses} />
                    )
                }
            </div >
        );
    }
}

const mapStateToProps = (state => ({
    expenses: getVisibleExpenses(state.expenses, state.filters),
}));

// First parameter vaneko - as a props ma k diney second parameter ma vako component ma ...
// First parameter takes a subpart of store's state
// and makes it accessible to second component props ...
export default connect(mapStateToProps)(ExpenseList);

