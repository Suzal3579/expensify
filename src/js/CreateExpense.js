import React, { Component } from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export class CreateExpense extends Component {
    onSubmit = (expense) => {
        this.props.onSubmitFromProps(expense);
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                <h1>Create Expense</h1>
                <ExpenseForm onSubmit={this.onSubmit} />
            </div >
        );
    }
}

const mapDispatchToProps = dispatch => ({
    onSubmitFromProps: expense => dispatch(addExpense(expense)),
});

export default connect(undefined, mapDispatchToProps)(CreateExpense);
