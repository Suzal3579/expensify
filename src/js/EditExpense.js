import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Button } from 'semantic-ui-react';

import firebase, { database } from '../firebase/firebase';
import { removeExpense, editExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpense extends Component {
    onSubmit = (expense) => {
        // Updating the data ...
        const currentUserUid = firebase.auth().currentUser.uid;
        database.ref(`users/${currentUserUid}/expenses`)
            .child(expense.id)
            .update({
                description: expense.description,
                note: expense.note,
                amount: expense.amount,
                createdAt: expense.createdAt,
            })
            .then(() => {
                this.props.history.push('/');
            });
    }
    onClick = () => {
        // Removing the item from firebase ...
        const currentUserUid = firebase.auth().currentUser.uid;
        database.ref(`users/${currentUserUid}/expenses`)
            .child(this.props.match.params.id)
            .remove(() => {
                this.props.history.push('/');
            });
    }
    render() {
        return (
            <div>
                <div className="col-md-6 col-md-offset-3">
                    <Button
                        fluid
                        negative
                        onClick={this.onClick}
                    >
                        Remove
                    </Button>
                    <br /><br />
                </div>
                <ExpenseForm
                    id={this.props.match.params.id}
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find(expense => expense.id === props.match.params.id),
});
const mapDispatchToProps = dispatch => ({
    editExpense: (id, expense) => { dispatch(editExpense(id, expense)); },
    removeExpense: (id) => { dispatch(removeExpense(id)); },
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);
