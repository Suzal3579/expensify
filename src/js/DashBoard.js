import React from 'react';
import ExpenseList from './ExpenseList';

import ExpenseListFilters from './ExpenseListFilters';

/* eslint-disable */
const DashBoard = props => (
    <div>
        <h1>DashBoard</h1>
        <h3>{props.match.params.id}</h3>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);
/* eslint-enable */

export default DashBoard;
