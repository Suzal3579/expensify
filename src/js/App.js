import React from 'react';
import { BrowserRouter, Route, Switch } from 'react-router-dom';
import { Provider } from 'react-redux';

import Header from './Header';
import DashBoard from './DashBoard';
import CreateExpense from './CreateExpense';
import Help from './Help';
import NoMatch from './NoMatch';
import store from '../store/store';
import { addExpense } from '../actions/expenses';
import { setTextFilter } from '../actions/filters';

const storeInstance = store();

storeInstance.dispatch(addExpense({
    description: 'Suzal is getting good at this.',
    amount: 90000,
    note: 'Team Lead',
    createdAt: 234521,
}));
storeInstance.dispatch(addExpense({
    description: 'Pramesh is getting good at many things.',
    amount: 190000,
    note: 'CTO',
    createdAt: 102123,
}));
storeInstance.dispatch(setTextFilter('Pramesh'));
setTimeout(() => {
    storeInstance.dispatch(setTextFilter('Suzal'));
}, 2000);

const App = () => (
    <Provider store={storeInstance}>
        <BrowserRouter>
            <div>
                <Header />
                <Switch>
                    <Route exact path="/" component={DashBoard} />
                    <Route path="/dashboard/:id" component={DashBoard} />
                    <Route path="/createExpense" component={CreateExpense} />
                    <Route path="/help" component={Help} />
                    <Route component={NoMatch} />
                </Switch>
            </div>
        </BrowserRouter>
    </Provider>
);

export default App;
