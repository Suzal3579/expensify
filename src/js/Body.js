import React from 'react';
import { Route, Switch } from 'react-router-dom';
import { Sidebar, Segment, Menu } from 'semantic-ui-react';

import DashBoard from './DashBoard';
import CreateExpense from './CreateExpense';
import EditExpense from './EditExpense';
import Help from './Help';
import MenuItems from './MenuItems';
import NoMatch from './NoMatch';

const Body = props => (
    <Sidebar.Pushable as={Segment}>
        <Sidebar
            as={Menu}
            animation="scale down"
            width="thin"
            direction="right"
            visible={props.visible}
            icon="labeled"
            vertical
            inverted
        >
            <MenuItems />
        </Sidebar>
        <Sidebar.Pusher>
            <Segment basic>
                <div id="body">
                    <Switch>
                        <Route exact path="/" component={DashBoard} />
                        <Route path="/createExpense" component={CreateExpense} />
                        <Route path="/editExpense/:id" component={EditExpense} />
                        <Route path="/help" component={Help} />
                        <Route component={NoMatch} />
                    </Switch>
                </div>
            </Segment>
        </Sidebar.Pusher>
    </Sidebar.Pushable>
);

export default Body;
