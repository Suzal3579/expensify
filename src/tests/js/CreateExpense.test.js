import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";

import { CreateExpense } from "../../js/CreateExpense";

const expenses = [{
    id: 'one1',
    description: 'desc 1',
    note: 'Some note 1',
    amount: 200,
    createdAt: 0,
}, {
    id: 'two2',
    description: 'Some 2',
    note: 'Some note 2',
    amount: 1200,
    createdAt: 5,
}, {
    id: 'three3',
    description: 'Some desc 3',
    note: 'Some note 3',
    amount: 3200,
    createdAt: 55,
}];

let onSubmitSpy, history, wrapper;
beforeEach(() => {
    onSubmitSpy = jest.fn();
    history = { push: jest.fn() }
    wrapper = shallow(<CreateExpense history={history} onSubmitFromProps={onSubmitSpy} />);
});

test("should render add expense page correctly", () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test("should render add expense page correctly", () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(onSubmitSpy).toHaveBeenLastCalledWith(expenses[1]);
}); 