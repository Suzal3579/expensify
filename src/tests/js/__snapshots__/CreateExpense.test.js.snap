// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render add expense page correctly 1`] = `
<div>
  <h1>
    Create Expense
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
